### PR TITLE
chore: allow crater runs

### LINKED ISSUE #6924: Add Zebra crate versions to dev-dependencies, to enable crater tests
## Motivation

If we add versions to all of Zebra's dev-dependencies, our crates will get tested by Rust compiler team `crater` runs.

This helps prevent build and test failures caused by future Rust compiler versions.

### Specifications

https://rustc-dev-guide.rust-lang.org/tests/crater.html

Also possibly the restrictions here, if they are still current:
https://www.reddit.com/r/rust/comments/f5xpib/psa_denywarnings_is_actively_harmful/

And we might need optional feature dependencies:
```toml
feature_name = ["optional_dep?/feature_if_dep_activated_by_another_feature"]
```

### Changes

- [ ] Add crate versions to all dev-dependencies on Zebra crates
    - [ ] Remove the tower-batch-control -> zebra-consensus circular dependency by copying those verifiers into tower-batch-control (and copying the tests into zebra-consensus so we're also testing the latest verifier code)
- [ ] Make sure the tests that are on by default:
    - run reasonably fast
    - don't use too much network
    - (check crater docs for any other requirements)

### Testing

This needs testing using `cargo release`, to make sure there aren't any other circular dependencies.

## Related Work

Split off PR #6908 and ticket #6149.

### PR DESCRIPTION
**Pull request by bishopcheckmate**
## Motivation

Fixes #6924

#### Check before marking the PR as ready for review:
  - [x] Will the PR name make sense to users?
  - [x] Does the PR have a priority label?
  - [x] Have you added or updated tests?
  - [x] Is the documentation up to date?

##### For significant changes:
  - [x] Is there a summary in the CHANGELOG?
  - [x] Can these changes be split into multiple PRs?

### Specifications

This change adds versions to all dev-dependencies so that they end up being packed together with crates on crates.io thus allowing crater runs.

It also removes the circular dependency between `tower-batch-control` and `zebra-consensus` by copying the Ed25519 verifier code into batch-control tests. Also updates the verifier tests in consensus to their state from tower-batch-control.

There are still unadressed points in #6924:

> Make sure the tests that are on by default:
> - run reasonably fast
> - don't use too much network
> - (check crater docs for any other requirements)

It's hard for me to define reasonably fast and too much networking. I can try to measure those on my PC if you want.
Regarding crater docs for any other requirements, I couldn't find any guidelines on that.

### Testing

This should be tested using `cargo publish`, however it's not currently possible due to git dependencies added in #8136.

### Reviewer Checklist

Check before approving the PR:
  - [ ] Does the PR scope match the ticket?
  - [ ] Are there enough tests to make sure it works? Do the tests cover the PR motivation?
  - [ ] Are all the PR blockers dealt with?
        PR blockers can be dealt with in new tickets or PRs.

_And check the PR Author checklist is complete._

### CODE REVIEW

File: tower-batch-control/tests/ed25519.rs
Comment (arya2) [inline tower-batch-control/tests/ed25519.rs:98]: Good idea defining the service here for testing. 

Optional: Since `zebra-consensus` has the same coverage now, I think it would be good to simplify this test service to remove the dependencies on `zebra-chain` and `metrics`. Let me know if you'd like me to open a suggestion/follow-up PR for this change.

**Review by arya2 — DISMISSED** [review]
This looks great, thank you!

I left an optional suggestion for removing the dependency on zebra-chain and metrics.

> Make sure the tests that are on by default:
> 
> run reasonably fast
> don't use too much network
> (check crater docs for any other requirements)

I didn't find any guidelines either, I'm not sure if there are any constraints in the code, or if it's manually observed and configured on a per-project basis here: https://github.com/rust-lang/crater/blob/master/config.toml

File: tower-batch-control/tests/ed25519.rs
Comment (bishopcheckmate) [reply tower-batch-control/tests/ed25519.rs:98]: Thanks,  I'll check

Comment (bishopcheckmate) [reply tower-batch-control/tests/ed25519.rs:98]: I've removed the metrics. with zebra-chain tho, would you like me to also replicate batch verifier in the batch control tests or something else?

Comment (arya2) [reply tower-batch-control/tests/ed25519.rs:98]: > with zebra-chain tho, would you like me to also replicate batch verifier in the batch control tests or something else?

I was thinking of changing `Item` to be a `Result`, so the test verifier looks for the first item that's an error and only tests the `Batch` service.

This PR looks good to merge if you'd like to skip it.

**Review by arya2 — DISMISSED** [review]

File: tower-batch-control/tests/ed25519.rs
Comment (bishopcheckmate) [reply tower-batch-control/tests/ed25519.rs:98]: I've just pushed the commit which uses just `ed25519` and not the whole `zebra-chain`

Comment (bishopcheckmate) [reply tower-batch-control/tests/ed25519.rs:98]: > This PR looks good to merge if you'd like to skip it.

> only tests the Batch service

I think I see the idea and I'd like to do that but I need to get the better understanding of the service to come up with some nice simple examples. I am not sure I'll have time for it before the weekend so we can merge it and I'll do a follow up if that's okay

**Review by arya2 — DISMISSED** [review]
This looks excellent, thank you!

File: tower-batch-control/tests/ed25519.rs
Comment (arya2) [reply tower-batch-control/tests/ed25519.rs:98]: > I think I see the idea and I'd like to do that but I need to get the better understanding of the service to come up with some nice simple examples. I am not sure I'll have time for it before the weekend so we can merge it and I'll do a follow up if that's okay

That sounds great, this PR has already gone beyond what was needed to close the issue (thank you).

I can open the follow up PR for your review if you'd like, I think it'll be easier to understand the service and the suggestion that way.

Comment (bishopcheckmate) [reply tower-batch-control/tests/ed25519.rs:98]: If you don't mind I like the harder path, it gives more project understanding

**Review by arya2 — DISMISSED** [review]

**Comment by arya2** [comment]
The CI failure seems unrelated to the changes, we should try to fix that or merge this manually.

**Review by arya2 — DISMISSED** [review]

**Comment by oxarbitrage** [comment]
@bishopcheckmate just FYI: your code is approved to be part of the zebra main branch but we are having CI issues with external pull requests. Right now, @gustavovalverde is working on fixing this issue. We will keep you posted, thanks for your contribution. I hope this problem do not discourage you to submit more code.

**Comment by bishopcheckmate** [comment]
hey, no problem at all. I'll definitely keep contributing here and there, just been busy recently

File: zebra-scan/Cargo.toml
Comment (arya2) [inline zebra-scan/Cargo.toml]: I resolved the merge conflict incorrectly here.

```suggestion
```

**Review by arya2 — DISMISSED** [review]

File: tower-batch-control/Cargo.toml
Comment (oxarbitrage) [inline tower-batch-control/Cargo.toml]:
```suggestion
tower-fallback = { path = "../tower-fallback/", version = "0.2.41-beta.10" }
```

File: tower-batch-control/Cargo.toml
Comment (oxarbitrage) [inline tower-batch-control/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

File: tower-fallback/Cargo.toml
Comment (oxarbitrage) [inline tower-fallback/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

File: zebra-chain/Cargo.toml
Comment (oxarbitrage) [inline zebra-chain/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

File: zebra-consensus/Cargo.toml
Comment (oxarbitrage) [inline zebra-consensus/Cargo.toml]:
```suggestion
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

File: zebra-rpc/Cargo.toml
Comment (oxarbitrage) [inline zebra-rpc/Cargo.toml]:
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-network = { path = "../zebra-network", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
```

File: zebra-rpc/Cargo.toml
Comment (oxarbitrage) [inline zebra-rpc/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
```

File: zebra-scan/Cargo.toml
Comment (oxarbitrage) [inline zebra-scan/Cargo.toml]:
```suggestion
```

File: zebra-script/Cargo.toml
Comment (oxarbitrage) [inline zebra-script/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
```

File: zebra-state/Cargo.toml
Comment (oxarbitrage) [inline zebra-state/Cargo.toml]:
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-test = { path = "../zebra-test/", version = "1.0.0-beta.34" }
```

File: zebrad/Cargo.toml
Comment (oxarbitrage) [inline zebrad/Cargo.toml]:
```suggestion
zebra-chain = { path = "../zebra-chain", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-consensus = { path = "../zebra-consensus", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-network = { path = "../zebra-network", version = "1.0.0-beta.34", features = ["proptest-impl"] }
zebra-scan = { path = "../zebra-scan", version = "0.1.0-alpha.3", features = ["proptest-impl"] }
zebra-state = { path = "../zebra-state", version = "1.0.0-beta.34", features = ["proptest-impl"] }
```

File: zebrad/Cargo.toml
Comment (oxarbitrage) [inline zebrad/Cargo.toml]:
```suggestion
zebra-node-services = { path = "../zebra-node-services", version = "1.0.0-beta.34", features = ["rpc-client"] }
```

File: zebrad/Cargo.toml
Comment (oxarbitrage) [inline zebrad/Cargo.toml]:
```suggestion
zebra-test = { path = "../zebra-test" version = "1.0.0-beta.34" }
zebra-grpc = { path = "../zebra-grpc" version = "0.1.0-alpha.1" }
```

File: zebrad/Cargo.toml
Comment (oxarbitrage) [inline zebrad/Cargo.toml]:
```suggestion
zebra-utils = { path = "../zebra-utils", version = "1.0.0-beta.34" }
```

**Review by oxarbitrage — COMMENTED** [review]
Bumping all the versions to the ones in our current release.

File: zebrad/Cargo.toml
Comment (oxarbitrage) [inline zebrad/Cargo.toml]: missed commas, sorry.

```suggestion
zebra-test = { path = "../zebra-test", version = "1.0.0-beta.34" }
zebra-grpc = { path = "../zebra-grpc", version = "0.1.0-alpha.1" }
```

**Review by oxarbitrage — APPROVED** [review]

**Pull request merged**
